api/v1/controllers/samples.js

```diff
@@ -99,26 +99,26 @@ function validateNonRunningCollectors(req) {
 
     const param = { key: { value: req.headers.TokenName } };
     u.findByKey(generators, param)
-    .then((generator) => {
-      if (!generator.currentCollector) {
-        const noCurrentCollector = 'Cannot accept Generator ' +
-          generator.name + ' samples without current collector';
-        return reject(
-          new apiErrors.ForbiddenError(
-            { explanation: noCurrentCollector, }
-          ));
-      } else if (generator.currentCollector &&
-        generator.currentCollector.status !== 'Running') {
-        const notRunning = 'Cannot accept samples from Collector ' +
+      .then((generator) => {
+        if (!generator.currentCollector) {
+          const noCurrentCollector = 'Cannot accept Generator ' +
+            generator.name + ' samples without current collector';
+          return reject(
+            new apiErrors.ForbiddenError(
+              { explanation: noCurrentCollector, }
+            ));
+        } else if (generator.currentCollector &&
+          generator.currentCollector.status !== 'Running') {
+          const notRunning = 'Cannot accept samples from Collector ' +
             generator.currentCollector.name + ' with status "' +
             generator.currentCollector.status + '"';
-        return reject(new apiErrors.ForbiddenError(
+          return reject(new apiErrors.ForbiddenError(
             { explanation: notRunning, })
           );
-      }
+        }
 
-      return resolve();
-    });
+        return resolve();
+      }).catch((err) => reject(err));
```

Comment:
What is the value of this reject? Isn't it the exact same thing as not catching, i.e. the promise rejection bubbles up to caller and its caller etc. until it *is* caught and handled?

Reply:
Sorry, I'm still not understanding. You replied "The caller is doing the .catch ... If I don’t do a catch and re throw the Error, there will be an unhandled exception." -- but if the caller is catching it, why would it be unhandled?

Reply:
I got your point, there is no point catching and rethrowing because the caller is doing that. Actually that was my first approach, the caller was responsable. After testing and forcing the resourceNotFoundException I saw this UnhandledPromise exception and the caller .catch was not happening.

1) If I remove the catch and re-throw, seems that the caller .catch is not getting properly.

<img width="1236" alt="screen shot 2018-08-23 at 18 32 42" src="https://user-images.githubusercontent.com/1087469/44541858-27d7e180-a703-11e8-8461-5debbb7ba892.png">

The exception came from verbs/util/findByName -> ResourcenotFoundException.

2) With catch and re-throw the caller gets the .catch
<img width="641" alt="screen shot 2018-08-23 at 18 38 51" src="https://user-images.githubusercontent.com/1087469/44542231-18a56380-a704-11e8-92fa-19a0cfabd85a.png">

Reply:
Fascinating. Please add a comment in the code indicating why this catch+rethrow is necessary.

Reply:
Yeap, ...

ok